feat(gatsby): Experimental unified routes

blainekasten opened the pull request:
## Description

Implementation of #24463.

This PR adds a few primary things:

- Allows collection routes from graphql data to be built using `{}` syntax in the filepath
- Allow client side dynamic routes using `[]` syntax in the filepath.
- Improvements to pass browser params to components using `props.params`
- path graphql field for models that use collection routes

### Overview

I'll give a rough overview of how this works now. First I'll explain the easier of the APIs which is the client side routes. 

### Client Side Routes

- In gatsby-plugin-page-creator, we have a [check](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-6cbc39b59dae5b8b2be93d18462f2648R44-R48) to see if a filepath includes `[]` in the path. This indicates that it is a client side path.
- Then we use gatsby-node to [create a page](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-443724ba2b2152744c2665f13339ed4c) with a correlated matchPath. The following examples are how it [translates](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-b9d6234b8752cb87b2dedb11b1989983R5-R8):
  - `src/pages/foo/[bar].js => /foo/:bar`
  - `src/pages/foo/[...bar].js => /foo/*bar`
  - `src/pages/foo/[bar]/baz/[...].js => /foo/:bar/baz/*`
- Then in the runtime, we [extract the params from the URL](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-917ba78a52f29b1a1fe42be34d81fc83R207-R222) for the dynamic segments and pass them to the component under `props.params`

### Collection Routes

This one is a bit more complicated, but it's because it's a richer feature 😄 

- in gatsby-plugin-page-creator we have a [check](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-6cbc39b59dae5b8b2be93d18462f2648R39-R40) to see if a filepath is using the `{}` in it's filepath to denote collection route.

- first we [validate the path is structured correctly](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-70ae45a17bf8a937ca5ebcef187a7f31R20), if not we watch the file to try again when the user fixes it.

- then we [extract a valid graphql query from the file path and an optional `unstable__collectionGraphql` function](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-70ae45a17bf8a937ca5ebcef187a7f31R28)

- Once we properly query for the data, we then get to [loop through each node and create the page](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-70ae45a17bf8a937ca5ebcef187a7f31R72-R93). On each loop we:
  - create the path from the file path and the data. e.g. if the path is `src/pages/post/{id}.js` and this node has an id of `"abc123"` then we would create the path as `/post/abc123`
  - We [extract](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-70ae45a17bf8a937ca5ebcef187a7f31R76) the params for the component from the data and the node and pass that to the component ultimately through this special context.__params key
  - We get the [context parameters](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-70ae45a17bf8a937ca5ebcef187a7f31R78) that will feed the page query and pass it on context too to reuse existing systems.
  - We also support an [optional matchPath](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-70ae45a17bf8a937ca5ebcef187a7f31R80) in case the user creates something like `src/pages/{id}/[page].js`
- [Lastly](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-70ae45a17bf8a937ca5ebcef187a7f31R95-R97) we add a [custom watcher](https://github.com/gatsbyjs/gatsby/pull/25204/files#diff-d72916b4bbf48b0a9f0c202b37e00991R10) for file changes to see if the query for generating the pages changes. If it does, we delete all the old pages and re-run the page creator for that file.


### Documentation

I'm going to create a separate PR to start documentation updates. The plan is for me to add a new tutorial page, and update the routing page. From there @AishaBlake and I will work together to complete all other updates. This feature affects _a lot_ of documentation

## Related Issues

#24463 

### Follow up Tickets in the future:

- [ ] First Class with incremental builds
- [ ] Redirect support in unified routes
- [x] named routing and linking
- [x] telemetry

vladar commented on packages/gatsby-plugin-page-creator/src/get-match-path.ts:
```diff
@@ -0,0 +1,19 @@
+    matchPath: srcPagesPath
+      .replace(`[...`, `*`)
+      .replace(`[`, `:`)
+      .replace(`]`, ``)
```
`replace` only replaces the very first occurence. So this URL:

```
/products/[:id]/[...page]/
```

Is transformed to:
```
 /products/:id/*page]
```

It probably requires a regex in this case.

blainekasten replied:
Good catch, I got a fix for this! Thanks

vladar commented on packages/gatsby-plugin-page-creator/src/derive-path.ts:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    // 3.d  replace the part of the slug with the actual value
+    path = path.replace(slugPart, value)
```
In my tests, it keeps the `.js` extension. Worked for me if changing this to

```js
pathWithoutExtension = pathWithoutExtension.replace(slugPart, value)
```

vladar commented on packages/gatsby-plugin-page-creator/src/extract-query.ts:
```diff
@@ -0,0 +1,88 @@
+): Record<string, string> {
+  const reversedParams = {}
+
+  absolutePath.split(`/`).forEach(part => {
```
Windows fix:

```suggestion
  absolutePath.split(/[\\/]/).forEach(part => {
```

vladar commented on packages/gatsby-plugin-page-creator/src/get-collection-route-params.ts:
```diff
@@ -0,0 +1,22 @@
+): Record<string, string> {
+  const params = {}
+  // remove the starting path to simplify the loop
+  const fileParts = filePath.split(`/`)
```
Windows fix:

```suggestion
  const fileParts = filePath.split(/[\\/]/)
```

vladar commented on packages/gatsby/src/utils/js-chunk-names.ts:
```diff
@@ -2,12 +2,40 @@ import { kebabCase } from "lodash"
+): string {
+  let newString = kebabedName
+
+  filePath.split(path.sep).forEach(part => {
```
`path.sep` works fine here but in other places I suggested using regex, so doing here for consistency too:

```suggestion
  filePath.split(/[\\/]/).forEach(part => {
```

Up to you :)

vladar commented on packages/gatsby-plugin-page-creator/src/get-collection-route-params.ts:
```diff
@@ -0,0 +1,24 @@
+  const params = {}
+  // remove the starting path to simplify the loop
+  const fileParts = filePath.split(path.sep)
+  const urlParts = urlPath.split(path.sep)
```
Windows-related: I always get `filePath` and `urlPath` with regular slashes here (e.g. `/posts/{fields__slug}/` vs. `\posts\{fields__slug}\` as you would expect on Windows). So `path.sep` doesn't do what expected here. 

It happens because those paths are produced by `createPath` utility here:

https://github.com/gatsbyjs/gatsby/blob/9720e82acc0afc462bd93f25f657d49b37f2da10/packages/gatsby-plugin-page-creator/src/create-pages-from-collection-builder.ts#L61-L63

And this utility actually always joins path segments by `/` using `path.posix.join`:

https://github.com/gatsbyjs/gatsby/blob/9720e82acc0afc462bd93f25f657d49b37f2da10/packages/gatsby-page-utils/src/create-path.js#L4-L9

So this method always returns empty `params` for me but surprisingly it doesn't seem to affect anything? Pages are still created and routing seems to be working. So what do we use those params for?

blainekasten replied:
Ah so it makes sense that this wouldn't _fail_. What it does is it extracts the url segments and puts it to the component on props.params. So I'm guessing for windows users, props.params in a collection route

blainekasten commented:
@vladar Thank you for your help! I fixed both of the reported bugs. `props.params` should not match all dynamic parts as you would expect using any other router. Also fixed the extract params bit so that props.params has the proper values on windows for  collection routes. I published a new canary if you don't mind testing again for me!

vladar commented on packages/gatsby/cache-dir/page-renderer.js:
```diff
@@ -37,6 +42,7 @@ class PageRenderer extends React.Component {
 }
 
 PageRenderer.propTypes = {
+  params: PropTypes.object.isRequired,
```
Do we need it now that `params` are actually added inside the `render` function and not passed in?

blainekasten replied:
removed!

vladar commented on packages/gatsby-plugin-page-creator/src/collection-extract-query-string.ts:
```diff
@@ -0,0 +1,54 @@
+  const modelType = /\{([a-zA-Z]+)\./.exec(absolutePath)?.[1]
+
+  if (!modelType) {
+    throw new Error(`You screwed up`)
```
I just got this error when testing 😅

Tested with the old naming convention, so basically, it failed with `{fields__slug}.js`. I feel that we need a message here explaining the expected name format.

blainekasten replied:
Haha yeah. I need to improve that. It was a quick thing :)

LekoArts commented on packages/gatsby-plugin-page-creator/src/__tests__/extract-query.ts:
```diff
@@ -0,0 +1,161 @@
+      ).toBe(`{allUnionQuery{nodes{id,parent{... on File{relativePath}}}}}`)
+    })
+
+    it(`supports nested graphql unions`, () => {
```
Isn't this is the exact same test as in line 103 ("supports graphql unions")?

What needs to also work is something like `/foo/bar/{Fruits.parent__(FruitsYaml)__parent__(File)__name}.js`, see:
- GitHub: https://github.com/LekoArts/unified-routes-demo/blob/master/src/pages/fruits/%7BFruits.parent__(FruitsYaml)__parent__(File)__name%7D.js
- Slack: https://gatsbyjs.slack.com/archives/C0137DT736V/p1595837372009500

blainekasten replied:
Yup, I messed up :) fixed!

pvdz commented on packages/gatsby-plugin-page-creator/src/extract-query.ts:
```diff
@@ -0,0 +1,103 @@
+  switchToPeriodDelimiters,
+} from "./path-utils"
+
+// Input queryStringParent could be a Modle or a full graphql query
```
Da heck is a Modle?

blainekasten replied:
fixed!

blainekasten commented:
@LekoArts there are e2e tests that verify that the `gatsbyPath` field works. I also added a bunch of validation to the field argument `filePath` to guide the user to get it right. Which there are unit tests for that too. So I think we're good on coverage, unless you have a specific request for test?

wardpeet commented on e2e-tests/development-runtime/package.json:
```diff
@@ -27,7 +27,7 @@
   "scripts": {
     "build": "gatsby build",
-    "develop": "cross-env CYPRESS_SUPPORT=y ENABLE_GATSBY_REFRESH_ENDPOINT=true gatsby develop",
+    "develop": "cross-env GATSBY_EXPERIMENTAL_ROUTING_APIS=1 CYPRESS_SUPPORT=y ENABLE_GATSBY_REFRESH_ENDPOINT=true gatsby develop",
```
can we create a new e2e test or do a double run until this is stable? When changing this test we lose track of the "stable" version of gatsby.

wardpeet commented on e2e-tests/production-runtime/package.json:
```diff
@@ -20,8 +20,8 @@
   "scripts": {
-    "build": "cross-env CYPRESS_SUPPORT=y gatsby build",
-    "build:offline": "cross-env TEST_PLUGIN_OFFLINE=y CYPRESS_SUPPORT=y gatsby build",
+    "build": "cross-env GATSBY_EXPERIMENTAL_ROUTING_APIS=1 CYPRESS_SUPPORT=y gatsby build",
+    "build:offline": "cross-env GATSBY_EXPERIMENTAL_ROUTING_APIS=1 TEST_PLUGIN_OFFLINE=y CYPRESS_SUPPORT=y gatsby build",
```
Same as develop runtime

wardpeet commented on e2e-tests/development-runtime/cypress/integration/unified-routing/collection-routing.js:
```diff
@@ -0,0 +1,22 @@
+    cy.visit(`/`).waitForRouteChange()
+  })
+
+  it(`can navigate to a collection route and see its content rendered`, () => {
+    cy.visit(`/collection-routing/root`).waitForRouteChange()
+
+    cy.getTestElement(`collection-routing-blog`)
+      .invoke(`attr`, `data-testslug`)
+      .then(slug => {
+        // should navigate us to an actual collection builder route.
+        cy.getTestElement(`collection-routing-blog`)
+          .first()
+          .click()
+          .waitForRouteChange()
+          .getTestElement(`slug`)
+          .invoke(`text`)
+          .should(`equal`, slug)
+      })
+  })
```
Do we want to add a few more test cases here? Like making sure all blog posts are rendered, ...

blainekasten replied:
Making sure the blog posts rendered is inferred. This test will crash if that doesn't work.

But added a test to ensure Unions work and that our collectionQuery api works too!

blainekasten commented on packages/gatsby-plugin-page-creator/src/path-utils.ts:
```diff
@@ -0,0 +1,84 @@
+}
+
+// Remove the file extension from the end of a path
+export function removeFileExtension(absolutePath: string): string {
+  return absolutePath.replace(/\.[a-z]+$/, ``)
+}
```
So turns out we can't use this actually. path.parse isn't too smart. It thinks that given `{Model.id}.js` that `.id}.js` is the extension
